src/storage/compression/roaring/compress.cpp

```diff
@@ -0,0 +1,487 @@
+	idx_t required_space = 0;
+	if (metadata.IsUncompressed()) {
+		// Account for the alignment we might need for this container
+		required_space += (AlignValue<idx_t>(reinterpret_cast<idx_t>(data_ptr))) - reinterpret_cast<idx_t>(data_ptr);
+	}
+	required_space += metadata.GetDataSizeInBytes(container_size);
+
+	idx_t runs_count = metadata_collection.GetRunContainerCount();
+	idx_t arrays_count = metadata_collection.GetArrayAndBitsetContainerCount();
+#ifdef DEBUG
+	idx_t current_size = metadata_collection.GetMetadataSize(runs_count + arrays_count, runs_count, arrays_count);
+	(void)current_size;
+	D_ASSERT(required_space + current_size <= GetRemainingSpace());
+#endif
+	if (metadata.IsRun()) {
+		runs_count++;
+	} else {
+		arrays_count++;
+	}
+	idx_t metadata_size = metadata_collection.GetMetadataSize(runs_count + arrays_count, runs_count, arrays_count);
+	required_space += metadata_size;
+
+	if (required_space > GetRemainingSpace()) {
+		return false;
+	}
+	return true;
+}
+
+void RoaringCompressState::InitializeContainer() {
+	if (total_count == analyze_state.total_count) {
+		// No more containers left
+		return;
+	}
+	auto container_index = GetContainerIndex();
+	D_ASSERT(container_index < container_metadata.size());
+	auto metadata = container_metadata[container_index];
+
+	idx_t container_size = AlignValue<idx_t, ValidityMask::BITS_PER_VALUE>(
+	    MinValue<idx_t>(analyze_state.total_count - container_state.appended_count, ROARING_CONTAINER_SIZE));
+	if (!CanStore(container_size, metadata)) {
+		idx_t row_start = current_segment->start + current_segment->count;
+		FlushSegment();
+		CreateEmptySegment(row_start);
+	}
+
+	// Override the pointer to write directly into the block
+	if (metadata.IsUncompressed()) {
+		data_ptr = reinterpret_cast<data_ptr_t>(AlignValue<idx_t>(reinterpret_cast<idx_t>(data_ptr)));
+		FastMemset(data_ptr, ~0, sizeof(validity_t) * (container_size / ValidityMask::BITS_PER_VALUE));
+		container_state.OverrideUncompressed(data_ptr);
+	} else if (metadata.IsRun()) {
+		auto number_of_runs = metadata.NumberOfRuns();
+		container_state.OverrideRun(data_ptr, number_of_runs);
+	} else {
+		auto cardinality = metadata.Cardinality();
+		container_state.OverrideArray(data_ptr, metadata.IsInverted(), cardinality);
+	}
+	data_ptr += metadata.GetDataSizeInBytes(container_size);
+	metadata_collection.AddMetadata(metadata);
+}
+
+//! FIXME: the real 'CreateTransientSegment' uses ValidityInitSegment which does unnecessary work (memset the whole
+//! block)
+unique_ptr<ColumnSegment> CreateTransientSegment(CompressionFunction &function, DatabaseInstance &db,
+                                                 const LogicalType &type, const idx_t start, const idx_t segment_size,
+                                                 const idx_t block_size) {
+	// Allocate a buffer for the uncompressed segment.
+	auto &buffer_manager = BufferManager::GetBufferManager(db);
```

Comment:
Can we move this to the ColumnSegment itself, e.g.:

```cpp
unique_ptr<ColumnSegment> ColumnSegment::CreateTransientSegment(DatabaseInstance &db, CompressionFunction &function, const LogicalType &type,
																const idx_t start, const idx_t segment_size,
																const idx_t block_size) {
	// Allocate a buffer for the uncompressed segment.
	auto &buffer_manager = BufferManager::GetBufferManager(db);
	auto block = buffer_manager.RegisterTransientMemory(segment_size, block_size);
	return make_uniq<ColumnSegment>(db, std::move(block), type, ColumnSegmentType::TRANSIENT, start, 0U, *function,
									BaseStatistics::CreateEmpty(type), INVALID_BLOCK, 0U, segment_size);

}

unique_ptr<ColumnSegment> ColumnSegment::CreateTransientSegment(DatabaseInstance &db, const LogicalType &type,
                                                                const idx_t start, const idx_t segment_size,
                                                                const idx_t block_size) {

	// Get the segment compression function.
	auto &config = DBConfig::GetConfig(db);
	auto function = config.GetCompressionFunction(CompressionType::COMPRESSION_UNCOMPRESSED, type.InternalType());
	return CreateTransientSegment(db, *uncompressed_function, ...);
}
```

Reply:
I've changed CreateTransientSegment, rather than make a separate overload
The only place where COMPRESSION_UNCOMPRESSED should actually be used is in `ColumnData::AppendTransientSegment`, where this is now explicit.